### PR TITLE
http/request: do not handle 304 Not Modified as a redirect code.

### PR DESCRIPTION
**Pull request by menakite**
If a `If-None-Match` or `If-Modified-Since` header is present in request and then a `304` response is received you would get a weird "missing location header for redirect" error from here: https://github.com/daurnimator/lua-http/blob/master/http/request.lua#L202

### CODE REVIEW

File: http/request.lua [hunk 1]
```diff
@@ -623,7 +623,7 @@ function request_methods:go(timeout)
 		self.cookie_store:store_from_request(request_headers, headers, self.host, self.site_for_cookies)
 	end
 
-	if self.follow_redirects and headers:get(":status"):sub(1,1) == "3" then
+	if self.follow_redirects and headers:get(":status"):sub(1,1) == "3" and not headers:get(":status") == "304" then
```
Comment (daurnimator) [inline http/request.lua:626]:
```suggestion
	if self.follow_redirects and headers:get(":status"):sub(1,1) == "3" and headers:get(":status") ~= "304" then
```

**Review by daurnimator — COMMENTED** [review]
I'd love a test case if you could; I'd insert one similar to https://github.com/daurnimator/lua-http/blob/ddab2835c583d45dec62680ca8d3cbde55e0bae6/spec/request_spec.lua#L175-L192

File: http/request.lua (hunk 1 above)
Comment (menakite) [reply http/request.lua]: Fixed in updated commit. Thanks.

**Comment by menakite** [comment]
> I'd love a test case if you could; I'd insert one similar to
> 
> https://github.com/daurnimator/lua-http/blob/ddab2835c583d45dec62680ca8d3cbde55e0bae6/spec/request_spec.lua#L175-L192

I have added a simple test case in the `:go` section, just after `follows redirects`.

I have to say these 2 tests were failing before and fail after this PR:
```
Failure → spec/request_spec.lua @ 887
http.request module :go method pays attention to HSTS
spec/request_spec.lua:946: Expected to be truthy, but value was:
(boolean) false

Failure → spec/request_spec.lua @ 1008
http.request module :go method handles HSTS corner case: 'preload' parameter
spec/request_spec.lua:1025: Expected to be truthy, but value was:
(boolean) false
```

**Comment by daurnimator** [comment]
> I have to say these 2 tests were failing before and fail after this PR:
> 
> ```
> Failure → spec/request_spec.lua @ 887
> http.request module :go method pays attention to HSTS
> spec/request_spec.lua:946: Expected to be truthy, but value was:
> (boolean) false
> 
> Failure → spec/request_spec.lua @ 1008
> http.request module :go method handles HSTS corner case: 'preload' parameter
> spec/request_spec.lua:1025: Expected to be truthy, but value was:
> (boolean) false
> ```

I'm not getting these failures when I run tests locally.
Are you overriding something/running a patched version or something?

**Comment by menakite** [comment]
Ah, if I understand correctly it tries to listen on port 443. If that's the case, I have a daemon running on that port, and then it obviously failed.

**Comment by daurnimator** [comment]
> Ah, if I understand correctly it tries to listen on port 443. If that's the case, I have a daemon running on that port, and then it obviously failed.

It passes port 0 to the kernel which allows the kernel to pick a free port. https://github.com/daurnimator/lua-http/blob/ee3cf4b4992479b8ebfb39b530694af3bbd1d1eb/spec/request_spec.lua#L892

I'd like to do a further investigation on it failing on your computer: could you open a new issue? Perhaps you could add some prints around https://github.com/daurnimator/lua-http/blob/ee3cf4b4992479b8ebfb39b530694af3bbd1d1eb/http/request.lua#L609 to see what path it is taking and if it gets to the `self.hsts:store` call

**Comment by menakite** [comment]
I believe it is an issue specific to Debian.

`self.hsts:store` is called, but it returns an error. I've tested with the following:
https://clbin.com/l0eUr

I have isolated the 2 failing tests and get:
```console
$ busted -c
self.hsts:store(self.host, sts) returned an error: max-age directive is required
◼self.hsts:store(self.host, sts) returned an error: max-age directive is required
◼
0 successes / 2 failures / 0 errors / 0 pending : 1.744904 seconds

Failure → spec/request_spec.lua @ 891
http.request module :go method pays attention to HSTS
spec/request_spec.lua:952: Expected to be truthy, but value was:
(boolean) false

Failure → spec/request_spec.lua @ 1018
http.request module :go method handles HSTS corner case: 'preload' parameter
spec/request_spec.lua:1035: Expected to be truthy, but value was:
(boolean) false
```

I think the problem is with the match on `sts_patt`.
Debian stable (12.7), but even testing and unstable, ships lpeg-patterns 0.4, which is pretty old:
https://packages.debian.org/search?keywords=lua-lpeg-patterns&searchon=names&exact=1&suite=all&section=all

I'm wondering if you can ping the Debian package maintainer?

Edit: affirmative, if I `git clone` the lpeg-patterns repository into the the lua-http directory (so that it has precedence), these tests succeed.
